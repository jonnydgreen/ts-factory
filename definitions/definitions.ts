import { BlockInput } from './block/block.type.ts';
import { FunctionDeclarationInput } from './function-declaration/function-declaration.type.ts';
import { IdentifierInput } from './identifier/identifier.type.ts';
import { KeywordTypeNodeInput } from './keyword-type-node/keyword-type-node.type.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { SourceFileInput } from './source-file/source-file.type.ts';

// Definition
export type Definition =
  | SourceFileInput
  | StatementInput
  | ExpressionInput
  | TypeNodeInput
  | ModifierInput;

// Member Name
export type MemberNameInput = IdentifierInput;

// Statement Inputs
export type StatementInput = FunctionDeclarationInput | BlockInput;

// Expression Inputs
export type ExpressionInput = IdentifierInput;

// Type Node Inputs
export type TypeNodeInput = KeywordTypeNodeInput;

// Input
export type Input = SourceFileInput;
