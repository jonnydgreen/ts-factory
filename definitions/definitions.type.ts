import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type {
  ExpressionInput,
  MemberNameInput,
  PropertyNameInput,
  StatementInput,
  TypeElementInput,
  TypeNodeInput,
} from './definitions.ts';
import { ModifierInput } from './modifier/modifier.type.ts';
import { TokenInput } from './token/token.type.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  T extends string | ts.PropertyName ? string | PropertyNameInput :
  T extends ts.Modifier ? ModifierInput :
  T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.TypeElement ? TypeElementInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  T extends ts.Token<ts.SyntaxKind> ? TokenInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;

export type PickDefinitionFields<T> = Omit<
  T,
  '__instructions' | 'leadingTrivia' | 'kind'
>;

export type DefinitionFields<T> = keyof PickDefinitionFields<T>;
