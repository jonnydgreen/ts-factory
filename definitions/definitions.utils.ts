import type { ts } from '../deps.ts';
import type { Instructions } from '../instructions/instructions.type.ts';
import type { ExpressionInput, MemberNameInput, StatementInput } from './definitions.ts';
import type { LeadingTriviaInput } from './trivia/trivia.type.ts';

export type MergeIntersections<I> = { [K in keyof I]: I[K] };

export type NonFnKeys<T> = {
  // deno-lint-ignore no-explicit-any
  [K in keyof T]: T[K] extends (...args: any[]) => any ? never : K;
}[keyof T];

// deno-fmt-ignore
export type MapTSTypes<T> =
  T extends ts.SyntaxKind ? T :
  T extends ts.NodeArray<infer TNode> ? MapTSTypes<TNode>[] :
  T extends ts.__String ? string :
  T extends string ? string :
  T extends ts.Statement ? StatementInput :
  T extends string | ts.MemberName ? string | MemberNameInput :
  // T extends ts.TypeNode ? TypeNodeInput :
  T extends ts.Expression ? ExpressionInput :
  T extends ts.SyntaxKind ? ExpressionInput :
  unknown;

export type CreateInput<
  TNode extends ts.Node,
  TNodeKeys extends NonFnKeys<TNode>,
  TPartialNodeKeys extends NonFnKeys<TNode> = TNodeKeys,
> = MergeIntersections<
  & {
    kind: TNode['kind'];
    leadingTrivia?: LeadingTriviaInput;
    __instructions?: Instructions;
  }
  & {
    -readonly [K in TNodeKeys]: MapTSTypes<Exclude<TNode[K], undefined>>;
  }
  & {
    -readonly [K in Exclude<TPartialNodeKeys, TNodeKeys>]?: MapTSTypes<
      Exclude<TNode[K], undefined>
    >;
  }
>;
