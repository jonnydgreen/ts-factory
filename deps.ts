// Path
import { join, resolve } from 'https://deno.land/std@0.147.0/path/mod.ts';
export const Path = {
  join,
  resolve,
};

// TS
export { ts } from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
export * as TSM from 'https://deno.land/x/ts_morph@15.1.0/mod.ts';
