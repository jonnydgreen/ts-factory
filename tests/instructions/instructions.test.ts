import { Definition } from '../../definitions/definitions.ts';
import { ts } from '../../deps.ts';
import {
  compileDefaultNodeInstructions,
  createPath,
  generateInstructions,
  processInstruction,
} from '../../instructions/instructions.ts';
import { Instruction, InstructionType } from '../../instructions/instructions.type.ts';
import { assertIsError, assertThrows, blocks } from '../../test.deps.ts';
import { AssertionError } from '../../utils/utils.assert.ts';
import { createSourceFile } from '../test-utils.ts';

blocks.describe('Instructions', () => {
  blocks.describe('generateInstructions', () => {
    blocks.it(
      'should throw an error if the definition kind does not match the node kind',
      () => {
        // Arrange
        const definition: Definition = {
          kind: ts.SyntaxKind.SourceFile,
          statements: [],
        };
        const sourceFile = createSourceFile('function hello() {}');
        const [statement] = sourceFile.getStatements();

        // Act
        const result = assertThrows(() => generateInstructions(statement, definition));

        // Assert
        assertIsError(
          result,
          AssertionError,
          'Definition of kind \'SourceFile\' does not match expected Node kind \'FunctionDeclaration\'',
        );
      },
    );

    blocks.it(
      'should throw an error if the definition contains an invalid field that we do not support',
      () => {
        // Arrange
        const definition: Definition = {
          kind: ts.SyntaxKind.SourceFile,
          statements: [],
          invalid: [{ kind: ts.SyntaxKind.AbstractKeyword }],
        } as unknown as Definition;
        const sourceFile = createSourceFile('function hello() {}');

        // Act
        const result = assertThrows(() => generateInstructions(sourceFile, definition));

        // Assert
        assertIsError(
          result,
          TypeError,
          'Unable to get field of name \'invalid\' from node of kind \'SourceFile\'',
        );
      },
    );
  });

  blocks.describe('processInstruction', () => {
    blocks.it('should error if invalid instruction type is passed', () => {
      // Arrange
      const sourceFile = createSourceFile();
      sourceFile.getKind = () => 98765;
      sourceFile.getKindName = () => 'InvalidKind';
      const instruction: Instruction = {
        type: InstructionType.ADD,
        path: createPath(),
        field: 'statements',
        definition: {
          kind: ts.SyntaxKind.FunctionDeclaration,
          name: 'hello',
          parameters: [],
        },
      };

      // Act
      const error = assertThrows(() => processInstruction(sourceFile, instruction));

      // Assert
      assertIsError(
        error,
        TypeError,
        'Unable to Process Instruction: unsupported parent node of kind InvalidKind',
      );
    });
  });

  blocks.describe('compileDefaultNodeInstructions', () => {
    blocks.it('should error if invalid instruction type is passed', () => {
      // Act
      const result = assertThrows(() =>
        compileDefaultNodeInstructions(
          createPath('path'),
          undefined,
          'field',
          'invalid-instruction-type' as InstructionType.SET,
        )
      );

      // Assert
      assertIsError(
        result,
        AssertionError,
        'Input invalid-instruction-type not supported',
      );
    });
  });
});
