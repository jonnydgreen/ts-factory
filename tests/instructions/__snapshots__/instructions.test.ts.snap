export const snapshot = {};

snapshot[`Instructions > ADD Instruction > should define an ADD instruction if the field is an array of nodes, the field nodes don't exist and no instructions are specified on the field definition 1`] = `
[
  {
    definition: {
      kind: "FunctionDeclaration",
      parameters: [
      ],
    },
    field: "statements",
    nodeID: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > ADD Instruction > should define an ADD instruction if the field is an array of nodes, the field nodes don't exist and instructions are specified on the field definition 1`] = `
[
  {
    definition: {
      kind: "FunctionDeclaration",
      parameters: [
      ],
    },
    field: "statements",
    nodeID: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > ADD Instruction > should define an ADD instruction if the field is an array of nodes, the field nodes exist and no instructions are specified on the field definition 1`] = `
[
  {
    definition: {
      kind: "FunctionDeclaration",
      parameters: [
      ],
    },
    field: "statements",
    nodeID: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > ADD Instruction > should define an ADD instruction if the field is an array of nodes and the field node is not found using the defined instruction ID 1`] = `
[
  {
    definition: {
      kind: "FunctionDeclaration",
      name: {
        kind: "Identifier",
        text: "hello",
      },
      parameters: [
      ],
    },
    field: "statements",
    nodeID: "",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > ADD Instruction > should not define an ADD instruction if the field is an array of nodes and the field node is found using the defined instruction ID 1`] = `
[
]
`;

snapshot[`Instructions > ADD Instruction > should define an ADD instruction if the field is an array of nodes and the field is a nested definition 1`] = `
[
  {
    definition: {
      kind: "ExportKeyword",
    },
    field: "modifiers",
    nodeID: "statements[0]",
    type: "ADD",
  },
  {
    definition: {
      kind: "AsyncKeyword",
    },
    field: "modifiers",
    nodeID: "statements[0]",
    type: "ADD",
  },
]
`;

snapshot[`Instructions > SET Instruction > should define an SET instruction if the field is a node, the field node doesn't exist and no instructions are specified on the field definition 1`] = `
[
  {
    definition: {
      kind: "VoidKeyword",
    },
    field: "type",
    nodeID: "statements[0]",
    type: "SET",
  },
]
`;

snapshot[`Instructions > SET Instruction > should define an SET instruction if the field is a node and the field node is not found using the defined instruction ID 1`] = `
[
  {
    definition: {
      kind: "VoidKeyword",
    },
    field: "type",
    nodeID: "statements[0]",
    type: "SET",
  },
]
`;

snapshot[`Instructions > SET Instruction > should not define an SET instruction if the field is a node, the field node exists and no instructions are specified on the field definition 1`] = `
[
]
`;

snapshot[`Instructions > SET Instruction > should not define an SET instruction if the field is a node and the field node is found using the defined instruction ID 1`] = `
[
]
`;

snapshot[`Instructions > INSERT Instruction > should define an INSERT instruction if the field is an array of nodes, a rule evaluates to an INSERT instruction and a numerical index is defined for the INSERT position 1`] = `
[
  {
    definition: {
      kind: "FunctionDeclaration",
      name: {
        kind: "Identifier",
        text: "hello",
      },
      parameters: [
      ],
    },
    field: "statements",
    index: 0,
    nodeID: "",
    type: "INSERT",
  },
]
`;

snapshot[`Instructions > INSERT Instruction > should define an INSERT instruction if the field is an array of nodes, a rule evaluates to an INSERT instruction and a string index is evaluated to an integer for the INSERT position 1`] = `
[
  {
    definition: {
      kind: "FunctionDeclaration",
      name: {
        kind: "Identifier",
        text: "hello",
      },
      parameters: [
      ],
    },
    field: "statements",
    index: 0,
    nodeID: "",
    type: "INSERT",
  },
]
`;

snapshot[`Instructions > INSERT Instruction > should not define an INSERT instruction if the field is an array of nodes and the rules does not evaluate to an instruction 1`] = `
[
]
`;
